src/map_core.rs

```diff
@@ -409,585 +104,291 @@ impl<K, V> Entries for IndexMapCore<K, V> {
-        debug_assert!(self.len() > 0);
-
-        let hash = entry.hash;
-        let actual_pos = ptrdistance(&self.entries[0], entry);
-        let probe = dispatch_32_vs_64!(self =>
-            find_existing_entry_at(&self.indices, hash, self.mask, actual_pos));
-        (probe, actual_pos)
+    /// Remove an entry by shifting all entries that follow it
+    pub(crate) fn shift_remove_full<Q>(&mut self, hash: HashValue, key: &Q) -> Option<(usize, K, V)>
+    where
+        Q: ?Sized + Equivalent<K>,
+    {
+        match self.find_equivalent(hash, key) {
+            Some(raw_bucket) => Some(self.shift_remove_bucket(raw_bucket)),
+            None => None,
+        }
     }
 
     /// Remove an entry by shifting all entries that follow it
-    pub(crate) fn shift_remove_found(&mut self, probe: usize, found: usize) -> (K, V) {
-        dispatch_32_vs_64!(self.shift_remove_found_impl(probe, found))
+    pub(crate) fn shift_remove_index(&mut self, index: usize) -> Option<(K, V)> {
+        let raw_bucket = match self.entries.get(index) {
+            Some(entry) => self.find_index(entry.hash, index).unwrap(),
+            None => return None,
+        };
+        let (_, key, value) = self.shift_remove_bucket(raw_bucket);
+        Some((key, value))
     }
 
-    fn shift_remove_found_impl<Sz>(&mut self, probe: usize, found: usize) -> (K, V)
-    where
-        Sz: Size,
-    {
-        // index `probe` and entry `found` is to be removed
+    /// Remove an entry by shifting all entries that follow it
+    fn shift_remove_bucket(&mut self, raw_bucket: RawBucket) -> (usize, K, V) {
         // use Vec::remove, but then we need to update the indices that point
         // to all of the other entries that have to move
-        self.indices[probe] = Pos::none();
-        let entry = self.entries.remove(found);
+        let index = unsafe {
+            self.indices.erase_no_drop(&raw_bucket);
+            raw_bucket.read()
+        };
+        let entry = self.entries.remove(index);
 
         // correct indices that point to the entries that followed the removed entry.
-        // use a heuristic between a full sweep vs. a `probe_loop!` for every shifted item.
-        if self.indices.len() < (self.entries.len() - found) * 2 {
-            // shift all indices greater than `found`
-            for pos in self.indices.iter_mut() {
-                if let Some((i, _)) = pos.resolve::<Sz>() {
-                    if i > found {
-                        // shift the index
-                        pos.set_pos::<Sz>(i - 1);
+        // use a heuristic between a full sweep vs. a `find()` for every shifted item.
+        let raw_capacity = self.indices.buckets();
+        let shifted_entries = &self.entries[index..];
+        if shifted_entries.len() > raw_capacity / 2 {
+            // shift all indices greater than `index`
+            unsafe {
+                for raw_bucket in self.indices.iter() {
```

Comment:
Ok, so the reason `RawTable::iter` is `unsafe` is because it has no lifetime parameter, it's a proxy for a bunch of raw pointers. However as long as we have the right borrowing behavior, there are no other traps or requirements on calling `next`, IIUC.

Reply:
`raw_bucket` here shadows the parameter `raw_bucket` in a way that obscures the logic, so I think it should use a different name. Same in the else branch.

Reply:
> However as long as we have the right borrowing behavior, there are no other traps or requirements on calling `next`, IIUC.

Yes, where the "right" behavior avoids anything that would invalidate pointers, namely reallocating.

> `raw_bucket` here shadows the parameter

OK, I'll rename these.